File: cpp/merge.cpp
#include <bits/stdc++.h>
#include <parallel/algorithm>

typedef std::vector<int> vec;
const int LEN = 100000000;

vec merge(const vec &arr1, const vec &arr2) {
    vec result(arr1.size() + arr2.size());
    unsigned i = 0, j = 0;
    while (i < arr1.size() && j < arr2.size()) {
         result[i+j] = arr1[i] < arr2[j] ? arr1[i++] : arr2[j++];
    }
    copy(arr1.begin()+i, arr1.end(), result.begin()+i+j);
    copy(arr2.begin()+j, arr2.end(), result.begin()+i+j);
    return result;
}

vec merge_sort(const vec &arr) {
    if (arr.size() <= 1) return arr;
    int mid = arr.size() / 2;
    vec left_half = merge_sort(vec(arr.begin(), arr.begin()+mid));
    vec right_half = merge_sort(vec(arr.begin()+mid, arr.end()));
    return merge(left_half, right_half);
}

void par_merge_sort(const vec& arr, vec& out) {
    if (arr.size() <= (1 << 13)) {
        out = merge_sort(arr);
    } else {
        int mid = arr.size() / 2;
        vec left_half, right_half;
        std::thread left(
            par_merge_sort, 
            vec(arr.begin(), arr.begin()+mid), 
            std::ref(left_half)
        );
        par_merge_sort(vec(arr.begin()+mid, arr.end()), right_half);
        left.join();
        out = merge(left_half, right_half);
    }
}

double bench(int size, std::function<void(vec&)> sorter) {
    srand(18);
    vec unsorted(size);
    timespec start, finish;
    std::generate(unsorted.begin(), unsorted.end(), rand);
    clock_gettime(CLOCK_MONOTONIC, &start);
    sorter(unsorted);
    clock_gettime(CLOCK_MONOTONIC, &finish);
    double elapsed = (finish.tv_sec - start.tv_sec) +
                     (finish.tv_nsec - start.tv_nsec) / 1e9;
    return elapsed;
}

int main(int argc, char **argv) {
    int len = argc == 1 ? LEN : atoi(argv[1]);
    std::cout << std::fixed << std::setprecision(9);
    double multi = bench(len, [](vec& v) {
        vec out;
        par_merge_sort(v, out);
    });
    double single = bench(len, [](vec& v) {
        merge_sort(v);
    });
    double builtin = bench(len, [](vec& v) {
        std::sort(v.begin(), v.end());
    });
    double builtin_multi = bench(len, [](vec& v) {
        __gnu_parallel::sort(v.begin(), v.end());
    });
    std::cout << "Multi-threaded:   " << multi         << "s\n" 
              << "Single-threaded:  " << single        << "s\n"
              << "Builtin sort:     " << builtin       << "s\n"
              << "Builtin parallel: " << builtin_multi << "s\n";
    return 0;
}

File: go/merge.go
package main
import (
    "fmt"
    "time"
    "math/rand"
    "os"
    "strconv"
    "sort"
)

func MergeSort(slice []int) []int {
	if len(slice) < 2 {
		return slice
	}
	mid := len(slice) / 2
	return Merge(
        MergeSort(slice[:mid]),
        MergeSort(slice[mid:]),
    )
}

func Merge(left, right []int) []int {
	size, i, j := len(left) + len(right), 0, 0
	slice := make([]int, size, size)
    for i < len(left) && j < len(right) {
        if left[i] < right[j] {
            slice[i+j] = left[i]
            i++
        } else {
            slice[i+j] = right[j]
            j++
        }
    }
    if i < len(left) {
        copy(slice[i+j:], left[i:])
    } else if j < len(right) {
        copy(slice[i+j:], right[j:])
    }
	return slice
}

func ParallelMergeSort(data []int, out chan []int) {
    if len(data) <= (1 << 13) {
        out <- MergeSort(data)
    } else {
        leftChan := make(chan []int)
        rightChan := make(chan []int)
        middle := len(data)/2
        go ParallelMergeSort(data[:middle], leftChan)
        go ParallelMergeSort(data[middle:], rightChan)
        ldata := <-leftChan
        rdata := <-rightChan
        close(leftChan)
        close(rightChan)
        out <- Merge(ldata, rdata)
    }
}

func Bench(N int, sorter func([]int)) time.Duration {
    r := rand.New(rand.NewSource(18))
    unsorted := make([]int, N)
    for i := 0; i < N; i++ {
        unsorted[i] = r.Intn(1 << 30)
    }
    now := time.Now()
    sorter(unsorted)
    elapsed := time.Now().Sub(now)
    return elapsed
}

func main() {
    N := 100000000
    if len(os.Args) > 1 {
        N, _ = strconv.Atoi(os.Args[1])
    }
    single := Bench(N, func (s []int) {
        MergeSort(s)
    })
    multi := Bench(N, func (s []int) {
        out := make(chan []int)
        go ParallelMergeSort(s, out)
        <-out
        close(out)
    })
    builtin := Bench(N, func (s []int) {
        sort.Ints(s)
    })
    fmt.Printf("Single thread: %v\n", single)
    fmt.Printf("Multi thread:  %v\n", multi)
    fmt.Printf("Builtin sort:  %v\n", builtin)
}
